File: apps/backend/internal/modules/workspace/animation_validator.go
package workspace

import (
	"encoding/json"
	"errors"
	"fmt"

	"github.com/Prodivix/prodivix/apps/backend/internal/platform/animationcontract"
)

var ErrAnimationValidationFailed = errors.New("Animation validation failed")

type animationMarker struct {
	ID                      string `json:"id"`
	RequiredInReducedMotion bool   `json:"requiredInReducedMotion"`
}

type animationReducedMotion struct {
	Kind       string `json:"kind"`
	TimelineID string `json:"timelineId"`
}

type animationTimeline struct {
	ID            string                 `json:"id"`
	MotionIntent  string                 `json:"motionIntent"`
	ReducedMotion animationReducedMotion `json:"reducedMotion"`
	Markers       []animationMarker      `json:"markers"`
}

type animationCompositionNode struct {
	ID            string                     `json:"id"`
	Kind          string                     `json:"kind"`
	TimelineID    string                     `json:"timelineId"`
	CompositionID string                     `json:"compositionId"`
	MarkerID      string                     `json:"markerId"`
	Children      []animationCompositionNode `json:"children"`
	Full          *animationCompositionNode  `json:"full"`
	Reduced       *animationCompositionNode  `json:"reduced"`
}

type animationComposition struct {
	ID          string                    `json:"id"`
	Root        animationCompositionNode  `json:"root"`
	ReducedRoot *animationCompositionNode `json:"reducedRoot"`
}

type animationDocument struct {
	Timelines          []animationTimeline    `json:"timelines"`
	Compositions       []animationComposition `json:"compositions"`
	EntryCompositionID string                 `json:"entryCompositionId"`
}

func validateAnimationDocument(payload json.RawMessage) error {
	if err := animationcontract.ValidateDocument(payload); err != nil {
		return animationValidationError(
			"/ must satisfy the generated Animation current schema: %v",
			err,
		)
	}
	var document animationDocument
	if err := json.Unmarshal(payload, &document); err != nil {
		return animationValidationError("/ could not be decoded: %v", err)
	}

	timelines := make(map[string]animationTimeline, len(document.Timelines))
	markerIDs := make(map[string]struct{})
	for index, timeline := range document.Timelines {
		if _, duplicate := timelines[timeline.ID]; duplicate {
			return animationValidationError(
				"/timelines/%d/id duplicates timeline %q",
				index,
				timeline.ID,
			)
		}
		timelines[timeline.ID] = timeline
		for _, marker := range timeline.Markers {
			markerIDs[marker.ID] = struct{}{}
		}
	}
	for index, timeline := range document.Timelines {
		path := fmt.Sprintf("/timelines/%d/reducedMotion", index)
		if timeline.MotionIntent == "essential" &&
			timeline.ReducedMotion.Kind != "retain" &&
			timeline.ReducedMotion.Kind != "timeline-ref" {
			return animationValidationError(
				"%s essential motion requires an explicit reduced variant",
				path,
			)
		}
		if timeline.MotionIntent == "continuous" &&
			timeline.ReducedMotion.Kind == "retain" {
			return animationValidationError(
				"%s continuous motion cannot be retained in reduced mode",
				path,
			)
		}
		if timeline.ReducedMotion.Kind != "timeline-ref" {
			continue
		}
		reduced, exists := timelines[timeline.ReducedMotion.TimelineID]
		if !exists || reduced.ID == timeline.ID {
			return animationValidationError(
				"%s/timelineId must reference a different existing timeline",
				path,
			)
		}
		required := requiredAnimationMarkerIDs(timeline.Markers)
		reducedRequired := requiredAnimationMarkerIDs(reduced.Markers)
		if len(required) != len(reducedRequired) {
			return animationValidationError(
				"%s must preserve required marker identity and order",
				path,
			)
		}
		for markerIndex := range required {
			if required[markerIndex] != reducedRequired[markerIndex] {
				return animationValidationError(
					"%s must preserve required marker identity and order",
					path,
				)
			}
		}
	}

	compositions := make(
		map[string]animationComposition,
		len(document.Compositions),
	)
	for index, composition := range document.Compositions {
		if _, duplicate := compositions[composition.ID]; duplicate {
			return animationValidationError(
				"/compositions/%d/id duplicates composition %q",
				index,
				composition.ID,
			)
		}
		compositions[composition.ID] = composition
	}
	if document.EntryCompositionID != "" {
		if _, exists := compositions[document.EntryCompositionID]; !exists {
			return animationValidationError(
				"/entryCompositionId references a missing composition",
			)
		}
	}

	nodeIDs := make(map[string]struct{})
	references := make(map[string][]string, len(compositions))
	for _, composition := range document.Compositions {
		roots := []*animationCompositionNode{&composition.Root}
		if composition.ReducedRoot != nil {
			roots = append(roots, composition.ReducedRoot)
		}
		for _, root := range roots {
			collectAnimationCompositionMarkers(root, markerIDs)
		}
		for _, root := range roots {
			if err := validateAnimationCompositionNode(
				root,
				composition.ID,
				timelines,
				compositions,
				markerIDs,
				nodeIDs,
				references,
			); err != nil {
				return err
			}
		}
	}
	if err := validateAnimationCompositionCycles(compositions, references); err != nil {
		return err
	}
	return nil
}

func requiredAnimationMarkerIDs(markers []animationMarker) []string {
	ids := make([]string, 0, len(markers))
	for _, marker := range markers {
		if marker.RequiredInReducedMotion {
			ids = append(ids, marker.ID)
		}
	}
	return ids
}

func collectAnimationCompositionMarkers(
	node *animationCompositionNode,
	markerIDs map[string]struct{},
) {
	if node.Kind == "marker" {
		markerIDs[node.MarkerID] = struct{}{}
	}
	for index := range node.Children {
		collectAnimationCompositionMarkers(&node.Children[index], markerIDs)
	}
	if node.Full != nil {
		collectAnimationCompositionMarkers(node.Full, markerIDs)
	}
	if node.Reduced != nil {
		collectAnimationCompositionMarkers(node.Reduced, markerIDs)
	}
}

func validateAnimationCompositionNode(
	node *animationCompositionNode,
	ownerID string,
	timelines map[string]animationTimeline,
	compositions map[string]animationComposition,
	markerIDs map[string]struct{},
	nodeIDs map[string]struct{},
	references map[string][]string,
) error {
	if _, duplicate := nodeIDs[node.ID]; duplicate {
		return animationValidationError(
			"/compositions contains duplicate node id %q",
			node.ID,
		)
	}
	nodeIDs[node.ID] = struct{}{}
	switch node.Kind {
	case "timeline-ref":
		if _, exists := timelines[node.TimelineID]; !exists {
			return animationValidationError(
				"/compositions node %q references missing timeline %q",
				node.ID,
				node.TimelineID,
			)
		}
	case "composition-ref":
		if _, exists := compositions[node.CompositionID]; !exists {
			return animationValidationError(
				"/compositions node %q references missing composition %q",
				node.ID,
				node.CompositionID,
			)
		}
		references[ownerID] = append(references[ownerID], node.CompositionID)
	case "marker":
		markerIDs[node.MarkerID] = struct{}{}
	case "settle":
		if node.MarkerID != "" {
			if _, exists := markerIDs[node.MarkerID]; !exists {
				return animationValidationError(
					"/compositions settle node %q references missing marker %q",
					node.ID,
					node.MarkerID,
				)
			}
		}
	}
	for index := range node.Children {
		if err := validateAnimationCompositionNode(
			&node.Children[index],
			ownerID,
			timelines,
			compositions,
			markerIDs,
			nodeIDs,
			references,
		); err != nil {
			return err
		}
	}
	if node.Full != nil {
		if err := validateAnimationCompositionNode(
			node.Full,
			ownerID,
			timelines,
			compositions,
			markerIDs,
			nodeIDs,
			references,
		); err != nil {
			return err
		}
	}
	if node.Reduced != nil {
		if err := validateAnimationCompositionNode(
			node.Reduced,
			ownerID,
			timelines,
			compositions,
			markerIDs,
			nodeIDs,
			references,
		); err != nil {
			return err
		}
	}
	return nil
}

func validateAnimationCompositionCycles(
	compositions map[string]animationComposition,
	references map[string][]string,
) error {
	visiting := make(map[string]bool, len(compositions))
	visited := make(map[string]bool, len(compositions))
	var visit func(string) error
	visit = func(id string) error {
		if visiting[id] {
			return animationValidationError(
				"/compositions contains a reference cycle through %q",
				id,
			)
		}
		if visited[id] {
			return nil
		}
		visiting[id] = true
		for _, reference := range references[id] {
			if err := visit(reference); err != nil {
				return err
			}
		}
		visiting[id] = false
		visited[id] = true
		return nil
	}
	for id := range compositions {
		if err := visit(id); err != nil {
			return err
		}
	}
	return nil
}

func animationValidationError(format string, args ...any) error {
	return fmt.Errorf("%w: %s", ErrAnimationValidationFailed, fmt.Sprintf(format, args...))
}

File: packages/animation/src/animationValidator.ts
import { sameCanonicalJson } from '@prodivix/shared/canonical';
import { isPlainObject } from '@prodivix/shared/safety';
import { normalizeAnimationDefinition } from './animationCodec';
import type {
  AnimationCompositionNode,
  AnimationDecodeIssue,
  AnimationDefinition,
} from './animation.types';

export const ANIMATION_VALIDATION_CODES = Object.freeze({
  documentInvalid: 'ANI_DOCUMENT_INVALID',
  targetInvalid: 'ANI_TARGET_INVALID',
  identityDuplicate: 'ANI_IDENTITY_DUPLICATE',
  referenceInvalid: 'ANI_REFERENCE_INVALID',
  compositionInvalid: 'ANI_COMPOSITION_INVALID',
  reducedMotionInvalid: 'ANI_REDUCED_MOTION_INVALID',
} as const);

export type AnimationValidationCode =
  (typeof ANIMATION_VALIDATION_CODES)[keyof typeof ANIMATION_VALIDATION_CODES];

export type AnimationValidationIssue = Readonly<{
  code: AnimationValidationCode;
  path: string;
  message: string;
}>;

export type AnimationValidationResult =
  | Readonly<{ valid: true; definition: AnimationDefinition; issues: [] }>
  | Readonly<{
      valid: false;
      issues: readonly AnimationValidationIssue[];
    }>;

const issue = (
  code: AnimationValidationCode,
  path: string,
  message: string
): AnimationValidationIssue => ({ code, path, message });

const collectCompositionNodes = (
  node: AnimationCompositionNode,
  path: string,
  visit: (node: AnimationCompositionNode, path: string) => void
): void => {
  visit(node, path);
  if (
    node.kind === 'sequence' ||
    node.kind === 'parallel' ||
    node.kind === 'stagger'
  ) {
    node.children.forEach((child, index) =>
      collectCompositionNodes(child, `${path}/children/${index}`, visit)
    );
  } else if (node.kind === 'conditional-variant') {
    collectCompositionNodes(node.full, `${path}/full`, visit);
    collectCompositionNodes(node.reduced, `${path}/reduced`, visit);
  }
};

const validateSemanticModel = (
  definition: AnimationDefinition
): AnimationValidationIssue[] => {
  const issues: AnimationValidationIssue[] = [];
  const timelineById = new Map(
    definition.timelines.map((timeline) => [timeline.id, timeline])
  );
  if (timelineById.size !== definition.timelines.length) {
    issues.push(
      issue(
        ANIMATION_VALIDATION_CODES.identityDuplicate,
        '/timelines',
        'Animation timeline identities must be unique.'
      )
    );
  }
  const compositionById = new Map(
    definition.compositions.map((composition) => [composition.id, composition])
  );
  if (compositionById.size !== definition.compositions.length) {
    issues.push(
      issue(
        ANIMATION_VALIDATION_CODES.identityDuplicate,
        '/compositions',
        'Animation composition identities must be unique.'
      )
    );
  }
  if (
    definition.entryCompositionId &&
    !compositionById.has(definition.entryCompositionId)
  ) {
    issues.push(
      issue(
        ANIMATION_VALIDATION_CODES.referenceInvalid,
        '/entryCompositionId',
        'Animation entryCompositionId must reference an existing composition.'
      )
    );
  }

  const markerIds = new Set<string>();
  definition.timelines.forEach((timeline) =>
    timeline.markers.forEach((marker) => markerIds.add(marker.id))
  );
  definition.compositions.forEach((composition, compositionIndex) => {
    const roots: readonly (readonly [AnimationCompositionNode, string])[] = [
      [composition.root, `/compositions/${compositionIndex}/root`],
      ...(composition.reducedRoot
        ? ([
            [
              composition.reducedRoot,
              `/compositions/${compositionIndex}/reducedRoot`,
            ],
          ] as const)
        : []),
    ];
    roots.forEach(([root, rootPath]) =>
      collectCompositionNodes(root, rootPath, (node) => {
        if (node.kind !== 'marker') return;
        markerIds.add(node.markerId);
      })
    );
  });

  definition.timelines.forEach((timeline, timelineIndex) => {
    const path = `/timelines/${timelineIndex}/reducedMotion`;
    if (timeline.motionIntent === 'essential') {
      if (
        timeline.reducedMotion.kind !== 'retain' &&
        timeline.reducedMotion.kind !== 'timeline-ref'
      ) {
        issues.push(
          issue(
            ANIMATION_VALIDATION_CODES.reducedMotionInvalid,
            path,
            'Essential motion requires an explicit retained or reduced timeline variant.'
          )
        );
      }
    }
    if (
      timeline.motionIntent === 'continuous' &&
      timeline.reducedMotion.kind === 'retain'
    ) {
      issues.push(
        issue(
          ANIMATION_VALIDATION_CODES.reducedMotionInvalid,
          path,
          'Continuous motion must stop or use a static reduced representation.'
        )
      );
    }
    if (timeline.reducedMotion.kind !== 'timeline-ref') return;
    const reducedTimeline = timelineById.get(timeline.reducedMotion.timelineId);
    if (!reducedTimeline || reducedTimeline.id === timeline.id) {
      issues.push(
        issue(
          ANIMATION_VALIDATION_CODES.referenceInvalid,
          `${path}/timelineId`,
          'Reduced motion timeline must reference a different existing timeline.'
        )
      );
      return;
    }
    const requiredMarkers = timeline.markers
      .filter((marker) => marker.requiredInReducedMotion)
      .map((marker) => marker.id);
    const reducedMarkers = reducedTimeline.markers
      .filter((marker) => marker.requiredInReducedMotion)
      .map((marker) => marker.id);
    if (!sameCanonicalJson(requiredMarkers, reducedMarkers)) {
      issues.push(
        issue(
          ANIMATION_VALIDATION_CODES.reducedMotionInvalid,
          path,
          'Reduced timeline variants must preserve required marker identity and order.'
        )
      );
    }
  });

  const nodeIds = new Set<string>();
  const compositionReferences = new Map<string, Set<string>>();
  definition.compositions.forEach((composition, compositionIndex) => {
    const references = new Set<string>();
    compositionReferences.set(composition.id, references);
    const roots: readonly (readonly [AnimationCompositionNode, string])[] = [
      [composition.root, `/compositions/${compositionIndex}/root`],
      ...(composition.reducedRoot
        ? ([
            [
              composition.reducedRoot,
              `/compositions/${compositionIndex}/reducedRoot`,
            ],
          ] as const)
        : []),
    ];
    roots.forEach(([root, rootPath]) =>
      collectCompositionNodes(root, rootPath, (node, path) => {
        if (nodeIds.has(node.id)) {
          issues.push(
            issue(
              ANIMATION_VALIDATION_CODES.identityDuplicate,
              `${path}/id`,
              `Animation composition node "${node.id}" is not document-unique.`
            )
          );
        }
        nodeIds.add(node.id);
        if (
          (node.kind === 'sequence' ||
            node.kind === 'parallel' ||
            node.kind === 'stagger') &&
          node.children.length === 0
        ) {
          issues.push(
            issue(
              ANIMATION_VALIDATION_CODES.compositionInvalid,
              `${path}/children`,
              'Animation composition containers require at least one child.'
            )
          );
        }
        if (
          node.kind === 'timeline-ref' &&
          !timelineById.has(node.timelineId)
        ) {
          issues.push(
            issue(
              ANIMATION_VALIDATION_CODES.referenceInvalid,
              `${path}/timelineId`,
              `Unknown Animation timeline "${node.timelineId}".`
            )
          );
        }
        if (node.kind === 'composition-ref') {
          references.add(node.compositionId);
          if (!compositionById.has(node.compositionId)) {
            issues.push(
              issue(
                ANIMATION_VALIDATION_CODES.referenceInvalid,
                `${path}/compositionId`,
                `Unknown Animation composition "${node.compositionId}".`
              )
            );
          }
        }
        if (
          node.kind === 'settle' &&
          node.markerId &&
          !markerIds.has(node.markerId)
        ) {
          issues.push(
            issue(
              ANIMATION_VALIDATION_CODES.referenceInvalid,
              `${path}/markerId`,
              `Unknown Animation settle marker "${node.markerId}".`
            )
          );
        }
      })
    );
  });

  const visiting = new Set<string>();
  const visited = new Set<string>();
  const visitComposition = (compositionId: string): void => {
    if (visiting.has(compositionId)) {
      issues.push(
        issue(
          ANIMATION_VALIDATION_CODES.compositionInvalid,
          '/compositions',
          `Animation composition reference cycle includes "${compositionId}".`
        )
      );
      return;
    }
    if (visited.has(compositionId)) return;
    visiting.add(compositionId);
    for (const reference of compositionReferences.get(compositionId) ?? []) {
      if (compositionById.has(reference)) visitComposition(reference);
    }
    visiting.delete(compositionId);
    visited.add(compositionId);
  };
  definition.compositions.forEach((composition) =>
    visitComposition(composition.id)
  );
  return issues;
};

/** Validates an unversioned canonical Animation current domain document. */
export const validateAnimationDefinition = (
  source: unknown
): AnimationValidationResult => {
  if (!isPlainObject(source)) {
    return {
      valid: false,
      issues: [
        issue(
          ANIMATION_VALIDATION_CODES.documentInvalid,
          '/',
          'Animation document content must be an object.'
        ),
      ],
    };
  }
  if (Object.hasOwn(source, 'version')) {
    return {
      valid: false,
      issues: [
        issue(
          ANIMATION_VALIDATION_CODES.documentInvalid,
          '/version',
          'Animation current domain documents do not contain wire versions.'
        ),
      ],
    };
  }
  const target = source.target;
  if (
    !isPlainObject(target) ||
    target.kind !== 'pir-document' ||
    typeof target.documentId !== 'string' ||
    !target.documentId.trim()
  ) {
    return {
      valid: false,
      issues: [
        issue(
          ANIMATION_VALIDATION_CODES.targetInvalid,
          '/target',
          'Animation documents require one explicit PIR document target.'
        ),
      ],
    };
  }
  const definition = normalizeAnimationDefinition(source);
  if (!definition || !sameCanonicalJson(source, definition)) {
    return {
      valid: false,
      issues: [
        issue(
          ANIMATION_VALIDATION_CODES.documentInvalid,
          '/',
          'Animation document content must already be canonical.'
        ),
      ],
    };
  }
  const issues = validateSemanticModel(definition);
  if (issues.length) return { valid: false, issues };
  return { valid: true, definition, issues: [] };
};

export type { AnimationDecodeIssue };
